asmcomp/power/selection.ml

```diff
@@ -43,13 +43,29 @@ let rec select_addr = function
   | exp ->
       (Alinear exp, 0, Debuginfo.none)
 
+let is_immediate n = n <= 0x7FFF && n >= -0x8000
+let is_immediate_logical n = n <= 0xFFFF && n >= 0
+
 (* Instruction selection *)
 
 class selector = object (self)
 
 inherit Selectgen.selector_generic as super
 
-method is_immediate n = (n <= 32767) && (n >= -32768)
+method is_immediate_test cmp n =
+  match cmp with
+  | Isigned _ -> is_immediate n
+  | Iunsigned _ -> is_immediate_logical n
+
+method is_immediate op n =
+  match op with
+  | Iadd | Imul -> is_immediate n
+  | Isub -> is_immediate (-n)  (* turned into add opposite *)
```

Comment:
I think this may also need to be changed in the RISCV backend, which also does the `-n` trick.

(I think that it would be less error-prone to pick `Iadd` during instruction selection, rather than having the assembly emitter convert between add and sub)

Reply:
> I think this may also need to be changed in the RISCV backend, which also does the -n trick.

Right.  I fixed that and tested on RISC-V.

> (I think that it would be less error-prone to pick Iadd during instruction selection, rather than having the assembly emitter convert between add and sub)

Maybe we've been through that already, but I like to leave the possibility open that a compilation pass will generate a Isub immediate after instruction selection.

Reply:
Fair enough!